overlord/devicestate/devicestate_test.go

```diff
@@ -2794,3 +2815,531 @@ func (s *deviceMgrSuite) TestDeviceCtxProvided(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(deviceCtx1, Equals, deviceCtx)
 }
+
+var snapYaml = `
+name: foo-gadget
+type: gadget
+`
+
+var gadgetYaml = `
+volumes:
+  pc:
+    bootloader: grub
+`
+
+func mockSnapWithData(c *C, yamlText string, si *snap.SideInfo, content map[string]string) *snap.Info {
+	info := snaptest.MockSnap(c, yamlText, si)
+	for where, what := range content {
+		path := filepath.Join(info.MountDir(), where)
+		err := os.MkdirAll(filepath.Dir(path), 0755)
+		c.Assert(err, IsNil)
+
+		err = ioutil.WriteFile(path, []byte(what), 0644)
+		c.Assert(err, IsNil)
+	}
+	return info
+}
+
+func (s *deviceMgrSuite) TestUpdateGadgetOnCoreSimple(c *C) {
+	var updateCalled bool
+	var passedRollbackDir string
+	restore := devicestate.MockGadgetUpdate(func(current, update *gadget.Info, path string) error {
+		updateCalled = true
+		passedRollbackDir = path
+		c.Assert(osutil.IsDirectory(path), Equals, true)
+		return nil
+	})
+	defer restore()
+	siCurrent := &snap.SideInfo{
```

Comment:
This block up to defer si.Unlock() looks the same between all tests - so maybe it can be extracted into a helper? Then each test would look something like: "setup mocked gadget udpate handler; boilderplate_helper(); check results". WDYT?